src/town_cmd.cpp

```diff
@@ -1199,6 +1199,72 @@ static bool GrowTownWithBridge(const Town *t, const TileIndex tile, const DiagDi
 	return false;
 }
 
+/**
+ * Grows the town with a tunnel.
+ *  First we check if a tunnel is reasonable.
+ *  If so we check if we are able to build it.
+ *
+ * @param t The current town
+ * @param tile The current tile
+ * @param tunnel_dir The valid direction in which to grow a tunnel
+ * @return true if a tunnel has been built, else false
+ */
+static bool GrowTownWithTunnel(const Town* t, const TileIndex tile, const DiagDirection tunnel_dir)
+{
+	assert(tunnel_dir < DIAGDIR_END);
+
+	Slope slope = GetTileSlope(tile);
+
+	/* Only consider building a tunnel if the starting tile is sloped properly. */
+	if (slope != InclinedSlope(tunnel_dir)) return false;
+
+	const int delta = TileOffsByDiagDir(tunnel_dir);
+	int max_tunnel_length = 0;
+
+	/* There are two conditions for building tunnels: Under a mountain and under an obstruction. */
+	if (CanRoadContinueIntoNextTile(t, tile, tunnel_dir)) {
+		/* Only tunnel under a mountain if the slope continues for at least 4 tiles, signifying a large mountain. */
+		TileIndex slope_tile = tile;
+		for (uint8 tiles = 0; tiles < 4; tiles++) {
+			slope = GetTileSlope(slope_tile);
+			if (slope != InclinedSlope(tunnel_dir) && !IsSteepSlope(slope) && !IsSlopeWithOneCornerRaised(slope)) return false;
+			slope_tile += delta;
+		}
+
+		/* For a mountain tunnel, the length limit starts with a base of 8 tiles, the minimum to reach the other side of a mountain with a 4-tile face.
+		* The allowable length increases by 1 for every 1000 town population.
```

Comment:
These kind of comments try to over-explain the situation, confusing people even more. The second sentence is very obvious from the code, so that for sure is not needed. The first sentence raises more questions than answers ... why does a 4-tile mountain require 8 tiles of tunnel? And why is there a 7 in the equation? I would personally avoid these kind of pitfalls, where you over-explain something, confusing the reader even more. Something simple as:

`/* More population means longer tunnels, but make sure we can at least cover a decent-sized mountain. */`

If the above comment works out, you can also simplify things. You can simply put the `min_tunnel_length` on 4 and do `) + min_tunnel_length + 3` or something, instead of `) + 7`.